File: 383. Ransom_Note/solution.cpp
class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        
        map<char,int> ran_map;
        map<char,int> mag_map;
        
        for(int i = 0 ; i < ransomNote.length() ; i++)
            ran_map[ransomNote[i]]++;
        
        for(int i = 0 ; i < magazine.length() ; i++)
            mag_map[magazine[i]]++;
        
        bool ans = true;
        for(auto it = ran_map.begin() ; it != ran_map.end() ; it++){
            if(mag_map.find(it->first) == mag_map.end() || mag_map[it->first] < it->second){
                ans = false;
                break;
            }
        }
        
        return ans;
        
    }
};

File: 383. Ransom_Note/solution.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        
        Map<Character,Integer> ran_map = new HashMap<Character,Integer>();
        Map<Character,Integer> mag_map = new HashMap<Character,Integer>();
        
        for(int i = 0 ; i < ransomNote.length() ; i++){
            if(ran_map.containsKey(ransomNote.charAt(i))){
                int old_val = ran_map.get(ransomNote.charAt(i));
                ran_map.replace(ransomNote.charAt(i), old_val + 1);
            }
            else
                ran_map.put(ransomNote.charAt(i), 1);
        }
        
        for(int i = 0 ; i < magazine.length() ; i++){
            if(mag_map.containsKey(magazine.charAt(i))){
                int old_val = mag_map.get(magazine.charAt(i));
                mag_map.replace(magazine.charAt(i), old_val + 1);
            }
            else
                mag_map.put(magazine.charAt(i), 1);
        }
        
        //System.out.println(mag_map.size());
        //System.out.println(ran_map.size());
        
        for(Map.Entry<Character,Integer> entry_now : ran_map.entrySet()){
            
            if(mag_map.containsKey(entry_now.getKey()) == false || entry_now.getValue() > mag_map.get(entry_now.getKey()) )
                return false;
        }
        
        return true;
    }
}